frmts/jp2kak/jp2kakdataset.cpp

```diff
@@ -2243,10 +2251,76 @@ JP2KAKCreateCopy( const char * pszFilename, GDALDataset *poSrcDS,
 
         CPLDebug("JP2KAK", "layer_bytes[] = %g\n",
                  static_cast<double>(layer_bytes[layer_count - 1]));
-    }
-    else
-    {
-        bReversible = true;
+    } else {
+        // Use RATE option.
+        char** dRateStr = CSLTokenizeStringComplex(CSLFetchNameValue(papszOptions,"RATE"),
+                                                   ",", FALSE, FALSE );
+        int rate_count = CSLCount( dRateStr );
+        if(rate_count <= 0) {
+            CPLError(CE_Fatal, CPLE_AppDefined,
+                     "RATE argument must be followed by a string identifying one or more bit-rates, separated by commas.\n");
+        } else if(bLayerOpt && (rate_count != layer_count || layer_count < rate_count) && rate_count > 2) {
+            // compare NUMBER of layers defined by LAYERS or CLayers and the number of layers defined by RATE
+            // and is not 1 or 2
+            //  If Clayers is not used, the number of layers is set to the number of rates specified here.
+            //  If Clayers is used to specify an actual number of quality layers, one of the following must be true:
+            //  1) the number of rates specified here is identical to the specified number of layers;
+            //  or 2) one or two rates are specified using this argument.
+            CPLError(CE_Fatal, CPLE_AppDefined,
+                     "The relationship between the number of bit-rates "
+                     "specified by the \"RATE\" argument and the number of quality layers "
+                     "explicitly specified via \"Clayers\" does not conform to the rules "
+                     "supplied in the description of the \"RATE\" argument\n");
```

Comment:
termination \n not needed
```suggestion
                     "supplied in the description of the \"RATE\" argument");
```